File: leetcode/Python/0004-h-median-of-two-sorted-arrays.py
"""
Binary Search
Time: O(log(m+n))
"""
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        total = len(nums1) + len(nums2)
        if total % 2 == 0:
            return (self.findKth(nums1, 0, nums2, 0, total // 2) +
                    self.findKth(nums1, 0, nums2, 0, total // 2 + 1)) / 2
        else:
            return self.findKth(nums1, 0, nums2, 0, total // 2 + 1)

    # k is not index, k >= 1
    def findKth(self, nums1, idx1, nums2, idx2, k):
        # nums1 is empty
        if idx1 >= len(nums1):
            return nums2[idx2 + k - 1]
        # nums2 is empty
        if idx2 >= len(nums2):
            return nums1[idx1 + k - 1]

        if k == 1:
            return min(nums1[idx1], nums2[idx2])

        key1 = float('inf')
        key2 = float('inf')

        if idx1 + k // 2 - 1 < len(nums1):
            key1 = nums1[idx1 + k // 2 - 1]

        if idx2 + k // 2 - 1 < len(nums2):
            key2 = nums2[idx2 + k // 2 - 1]

        if key1 < key2:
            return self.findKth(nums1, idx1 + k // 2, nums2, idx2, k - k // 2)
        else:
            return self.findKth(nums1, idx1, nums2, idx2 + k // 2, k - k // 2)

File: leetcode/C++/0004-h-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        const int m = nums1.size();
        const int n = nums2.size();
        int total = m + n;
        if (total & 0x1)
            return find_kth(nums1.begin(), m, nums2.begin(), n, total / 2 + 1);
        else
            return (find_kth(nums1.begin(), m, nums2.begin(), n, total / 2)
                    + find_kth(nums1.begin(), m, nums2.begin(), n, total / 2 + 1)) / 2.0;
    }
private:
    static int find_kth(vector<int>::const_iterator A, int m, 
                        vector<int>::const_iterator B, int n, int k) {
        // always assume that m is equal or smaller than n
        if (m > n)
            return find_kth(B, n, A, m, k);
        if (m == 0)
            return *(B + k - 1);
        if (k == 1)
            return min(*A, *B);
        
        // divide k into two parts
        int ia = min(k / 2, m), ib = k - ia;
        if (*(A + ia - 1) < *(B + ib - 1))
            return find_kth(A + ia, m - ia, B, n, k - ia);
        else if (*(A + ia - 1) > *(B + ib - 1))
            return find_kth(A, m, B + ib, n - ib, k - ib);
        else 
            return A[ia - 1];
    }
};